### PR TITLE
add cuda::Stream constructor with cuda stream flags

### PR DESCRIPTION
**Pull request by diablodale**
adds feature described in issue opencv/opencv#19285
includes code, test, and doc.

TLDR: adds a constructor for cuda::Stream which accepts a raw cudaStream_t and assumes ownership of it including destroying the raw CUDA stream in the Stream's destructor.

<cut/>

### Pull Request Readiness Checklist

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [X] The PR is proposed to proper branch
- [X] There is reference to original bug report and related work
- [X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [X] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
buildworker:Custom=linux-4
build_image:Custom=ubuntu-cuda:18.04
```

### CODE REVIEW

**Comment by diablodale** [comment]
updated PR to remove python wrap for the new constructor. Python and move semantics are not supported. And not workable for python to have a true cudaStream_t from OpenCV or numby. Might be possible via CuPy, but that is out of scope.

File: modules/core/include/opencv2/core/cuda.hpp [hunk 1]
```diff
@@ -656,6 +656,19 @@ class CV_EXPORTS_W Stream
...
+    @note The parameter must be a cudaStream_t. The parameter is void* to align
+    with cudaPtr(). The ownership of the cudaStream_t will move to this Stream.
+    The destructor of this Stream will cudaStreamDestroy(cudaStream).
+    @code
+        cudaStream_t cudaStream;
+        cudaStreamCreateWithFlags(&cudaStream, cudaStreamNonBlocking);
+        cv::cuda::Stream cvStream(std::move((void*)cudaStream));
+    @endcode
+     */
+    Stream(void*&& cudaStream);
```
Comment (alalek) [inline modules/core/include/opencv2/core/cuda.hpp:670]: It make sense to avoid "void*" ctors in public API (because there is no strong type check and many implicit conversions).

Consider using:
```
static Stream fromCudaStream(void*&& cudaStream);
```

Usage:

```.diff
-cv::cuda::Stream cvStream(std::move((void*)cudaStream));
+auto cvStream = cv::cuda::Stream::fromCudaStream(std::move((void*)cudaStream));
```

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

File: modules/core/include/opencv2/core/cuda.hpp (hunk 1 above)
Comment (diablodale) [reply modules/core/include/opencv2/core/cuda.hpp:670]: The ask is a constructor for the reasons that I wrote in the original issue https://github.com/opencv/opencv/issues/19285. Please reference there for details including the unusable existing function `cv::cuda::StreamAccessor::wrapStream()`. 

A constructor can provide the instantiated object itself or a pointer to an instantiated object. They are not the same. A pointer is raw and unsafe to use without further wrapping it with unique_ptr or shared_ptr. The object itself is scoped, RAII, etc. Technically both are valid. But only a constructor can create an object itself into an lvalue.

```c++
// safe, destructs when out of scope
cv::cuda::Stream stream_obj(std::move((void *)my_cudaStream_t));

// dangerous raw pointer, does not destruct itself, often multiple memory bugs by app programmers
cv::cuda::Stream* stream_ptr = cv::cuda::Stream::fromCudaStream(std::move((void *)my_cudaStream_t));

// safer, valid, more work
std::unique_ptr<cv::cuda::Stream> stream_ptr(cv::cuda::Stream::fromCudaStream(std::move((void *)my_cudaStream_t)));

// safe, simple, and requires a constructor
struct threadTLSData{
    cv::cuda::Stream cvStream;
    int otherThings;
};
return {
    cv::cuda::Stream(std::move((void *)_stream)),
    42
};

// without a constructor have to wrap it all and requires C++11 compiler correctly supports copy elision
struct threadTLSData{
    std::unique_ptr<cv::cuda::Stream> cvStream; // share_ptr isn't as strict with elision, but has more overhead when we don't need it here
    int otherThings;
};
return {
    // also can not use std::make_unique or std::make_shared because they both require a constructor
    // also make_unique is only in C++14 and newer yet OpenCV wants to target C++11
    // this direct unique_ptr constructor method is often unsafe in C++11 due to parameter evaluation order
    // with exceptions. For safety, need make_unique in C++14 -or- use C++17 standard that fixed this
    // https://www.learncpp.com/cpp-tutorial/stdunique_ptr/#urvanov-syntax-highlighter-600141f5f0789724071531
    std::unique_ptr<cv::cuda::Stream>(cv::cuda::Stream::fromCudaStream(std::move((void *)my_cudaStream_t))),
    42
};
```

OpenCV is often not strong type checked. As I discuss in the issue #19285, this is a development choice by the core team. They did not want to include headers (e.g. CUDA, OpenCL) and fix issues at compile time, so they wrote generic parameters/returns like `void*` and throw asserts at runtime. For example: `cv::ocl::Device(void*)`, `void* cv::cuda::Stream::cudaPtr()`, `cv::cuda::Stream::Stream()` that throws, or the whole auxiliary namespace `cv::cuda::StreamAccessor`.

I would have taken a different approach with strong type safety and fix at compile time.  However since the core team choose the generic `void*` and assert approach, I suggest we continue that approach for consistency plus the above benefits of having a constructor. I've coded it like that in the PR.

### Options

#### Constructor with `Stream::Stream(void*&& cudaStream)`

See original post and PR

#### Static factory `static Stream::fromCudaStream(void*&& cudaStream)`

I can write a lesser feature with a static factory like `static cv::cuda::Stream::fromCudaStream(void*&& cudaStream)`. It will have the limitations and more difficult usage that I describe above in words and pseudocode.

#### Use type-safety

Make a breaking change to the `cv::cuda` module to end the use of `void*` on public APIs. This requires the inclusion some cuda headers so that types like `cudaStream_t` can be used in APIs like this new constructor, existing `Stream::cudaPtr()`, etc.

#### Gated availability of type-safe constructor

It should be easy to use `#ifdef HAVE_CUDA` to test for availability of CUDA, optionally include CUDA headers and declare a type-safe constructor. This would give us a constructor, type-safe, and not required CUDA headers to be loaded unless OpenCV is being built with cuda. For example pseudocode

```
#ifdef HAVE_CUDA
    #include<cuda.h>
    cv::cuda::Stream::Stream(cudaStream_t&& cudaStream);
#endif
```

Comment (alalek) [reply modules/core/include/opencv2/core/cuda.hpp:670]: > cv::ocl::Device(void*)

Exactly in this area there are several issues with "void*" type in constructor.

> unique_ptr

Doesn't make sense, because this class is smart ptr itself.
Anyway you can always try to add this "safe" factory:

```
static std::unique_ptr<Stream> makeUniqueFromCudaStream(void*&& cudaStream);
```

> copy elision

Try to add move ctor.

---

For other proposals there is ABI compatibility (between configurations) with BUILD_CUDA_STUBS CMake option.

Comment (diablodale) [reply modules/core/include/opencv2/core/cuda.hpp:670]: I believe there is a possibility here. We just haven't found something that you will accept into the codebase and that meets the needs of users like me. What else should we explore?

> Exactly in this area there are several issues with "void*" type in constructor.

Help me understand. Are you writing as a core team member....that the core team *does not* want to have `void*` in public opencv APIs? I ask because `void* Stream::cudaPtr()` and three others was added 11 months ago on 26 February 2020. These `void*` uses are not old...they are new. Someone in the core team allowed it, here is the commit https://github.com/opencv/opencv/commit/53c0189ed7efe52ad65c38cac04181155e320ca8

It is difficult for contributors to understand how to contribute when there is conflicting information. You are telling me one thing (don't use void*) and only 11 months ago I see it was allowed and merged (with more `void*` use if we look older). My intention is not to complain or make the core team wrong. Instead, I want to contribute in a positive way. I need clear consistent strategy and implementation rules. If I don't have that, it is difficult for me to know how to reverse-engineer and write new code. 🤪

> unique_ptr

This is a important need. Static factories return *pointers* to objects; it returns `cuda::Stream *`. This is different than constructors. A constructor returns `cuda::Stream`. If we take the static factory approach, then it returns a pointer...therefore I want to make it safe. No memory leaks, ref count errors, double freeing, ownership, etc! 🔨 These are real pointer problems that happen in opencv (I've found and fixed some) and in apps that use opencv as a library. We are human, we make mistakes. I strongly prefer designs that prevent these mistakes.  C++11 and newer has good mechanisms to prevent them.

When adding a constructor, we get all the benefits of RAII and scope that I write above. If instead, we have a static factory that returns a pointer, then how do we make it safe? Yes, it could return a unique_ptr instead. That aligns with C++ core guideline 26 https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f26-use-a-unique_ptrt-to-transfer-ownership-where-a-pointer-is-needed

However, that doesn't align with the typical use of static factories like discussed in guideline https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c50-use-a-factory-function-if-you-need-virtual-behavior-during-initialization. Usually static factories are used for one of two reasons:

1. To return a pointer to a class object from a pool of reusable existing objects, or to return a singleton object.
2. To return a pointer to a subclass based on the parameters passed. It does not return a pointer to itself (the base object)

A static factory like below could technically work, but it "feels" strange, lesser, and not a good precedent to set.
```c+
static std::unique_ptr<cv::cuda::Stream> cv::cuda::Stream::makeUniqueFromCudaStream(void*&& cudaStream);
```

This feeling is so strong, I would suggest not doing it. Instead, overloading a function in that `cv::cuda::StreamAccessor` namespace where it can be strongly typed. Pseudocode...

```c++
Stream cv::cuda::StreamAccessor::wrapStream(cudaStream_t&& stream);
```

It is consistent with the already existing wrapStream() function. But, it is not consistent with the already existing `void*` use. Personally, I probably wouldn't use a function like this. It can't be used to initialize a member field; at best a field would be default constructed and then copy assigned from this function. I want a constructor for safety and simplicity.

> Try to add move ctor.

That is my original post and PR. It is a move constructor.

Comment (alalek) [reply modules/core/include/opencv2/core/cuda.hpp:670]: > "void*" type in constructor.
> does not want to have void* in public opencv APIs

The first only.

> Static factories return pointers to objects; it returns cuda::Stream *

This doesn't follow RAII.

---

> StreamAccessor

I believe this is the right way.
There is already some related strong-typed [code](https://github.com/opencv/opencv/blame/63bb2abadab875fc648a572faccafee134f06fc8/modules/core/src/cuda_stream.cpp#L563-L566) (added in C++98 epoch, so there is no move semantic).

Comment (alalek) [reply modules/core/include/opencv2/core/cuda.hpp:670]: /cc @nglee @cudawarped @tomoaki0705 Any through about this?

Comment (nglee) [reply modules/core/include/opencv2/core/cuda.hpp:670]: To the best of my understanding, `cudaStream_t` is itself a pointer.
```C++
// driver_types.h
typedef __device_builtin__ struct CUstream_st *cudaStream_t;
```

So I think it's not clear why we would need a constructor that takes ownership of a `cudaStream_t`. We can create a `cv::cuda::Stream` object and get the managed `cudaStream_t` object with `cv::cuda::StreamAccessor::getStream(const Stream& stream)`.

Comment (nglee) [reply modules/core/include/opencv2/core/cuda.hpp:670]: I think the constructor that takes ownership of a `cudaStream_t` would need to check whether the `cudaStream_t` passed as an argument is already initialized with `cudaStreamCreate()` or not. I can't find an API in CUDA that does that. So I think it is hard to implement one.

I'm not sure in which situation we would need such a constructor.

Also, `cudaStream_t` is itself a pointer, so I think there are not many benefits from `std::move`-ing it.
```C++
// driver_types.h
typedef __device_builtin__ struct CUstream_st *cudaStream_t;
```

Comment (diablodale) [reply modules/core/include/opencv2/core/cuda.hpp:670]: Move is the C++11 and newer semantic to declare and force a transfer of ownership. Without it, ownership is ambigious -- often duplicated, and leading to leaks and/or double-deletes.
After the move into the constructor I proposed...the previously owning variable has no data and no memory. I can't be used further in code. These are the good things that C++11 and newer has to help coders write better code. That is the value: clear ownership....less bugs...cleaner code.

I previously looked myself, CUDA doesn't have an API to validate a cudaStream_t. However, this is something I would not expect OpenCV to do. OpenCV doesn't validate the `data` pointer for cv::mat, cv::umat or any of many many other classes that have constructors accepting pointers.

Comment (nglee) [reply modules/core/include/opencv2/core/cuda.hpp:670]: I wanted to say that move semantics is not that useful for primitive types such as pointers.

https://stackoverflow.com/questions/14679605/do-built-in-types-have-move-semantics

I think just declaring the constructor as:
```C++
Stream(void* cudaStream);
```
and calling it as the following seems enough.
```C++
cudaStream_t cudaStream;
cudaStreamCreateWithFlags(&cudaStream, cudaStreamNonBlocking);
cv::cuda::Stream cvStream((void*)cudaStream));
```

However, speaking of other OpenCV constructors, those that accept pointers do not take ownership. I'm not sure if we need one for `Stream`.

Comment (diablodale) [reply modules/core/include/opencv2/core/cuda.hpp:670]: Snap! I have an alternative.

Thanks @nglee you gave the hint at the std::move doesn't actually do what I though it did. More research and I now see that std::move is very much only a cast and that it is the responsibility of the constructor/assignment operator to manually change memory. For example STL leaves the var "placed in a valid but unspecified state".

I could change my PR to set the `cudaStream_t` to null. Unfortunately, a null cudaStream_t (the pointer itself rather than to which it points) means the default CUDA stream. This would prevent double cudaStreamDestroy() because the destroy api only works on async streams and would likely return an error or silent ignore (I could test it). Is this any better? Ehhh...marginally. It does transfer ownership of `cudaStream_t`. But that means the previously owning var now has a value of null which a programmer could continue to use in errant code to do things without realizing it is now with the default CUDA stream.

The legacy of the `cudaStream_t` might be the death of this. I do not see any CUDA define for an invalid stream. (They have them for devices and cpus...but not streams). I have no value that I can set the previously owning var so that it can not be misused.

My main use case is to create a Stream with a specific type of cudaStream_t. I need unique per-thread non-default async streams. These can be created with `cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking)`. OpenCV aready has code with this approach at `cv::dnn::cuda4dnn::csl::UniqueStream()`

The prototype for that nvidia cuda API is: `cudaStreamCreateWithFlags(cudaStream_t *pStream, unsigned int flags)`

### Alternative

Have a Stream constructor like this

```c++
CV_WRAP cv::cuda::Stream::Stream(unsigned int cudaFlags);

// example use
cv::cuda::Stream mystream(cudaStreamNonBlocking);
```

Benefits are:

* It is a constructor with all its benefits
* No CUDA headers are needed in the public opencv API
* Ownership is always contained within the Stream()
* It can be used from Python

Comment (nglee) [reply modules/core/include/opencv2/core/cuda.hpp:670]: The alternative seems good to me. 👍

Comment (diablodale) [reply modules/core/include/opencv2/core/cuda.hpp]: I've updated branch and force pushed. It should arrive soon.

**Comment by diablodale** [comment]
I see two sets of warnings/errors in the CI builds. I request assist as I do not understand them

### warnings caused by the Py wrapper infrastructure

https://pullrequest.opencv.org/buildbot/builders/precommit_opencl_linux/builds/27820/steps/compile%20release/logs/warnings%20%282%29. The parameter on the C++ constructor is intentionally required and should not have a default value. This is the generated code.

```C++
{
    PyObject* pyobj_cudaFlags = NULL;
    unsigned cudaFlags;

    const char* keywords[] = { "cudaFlags", NULL };
    if( PyArg_ParseTupleAndKeywords(py_args, kw, "O:Stream", (char**)keywords, &pyobj_cudaFlags) &&
        pyopencv_to(pyobj_cudaFlags, cudaFlags, ArgInfo("cudaFlags", 0)) )
    {
        new (&(self->v)) Ptr<cv::cuda::Stream>(); // init Ptr with placement new
        if(self) ERRWRAP2(self->v.reset(new cv::cuda::Stream(cudaFlags)));
        return 0;
    }
}
```

### fail in the "custom" CI platform

I see errors in some .cu file that is likely generated by something that is part of a maybe a contrib module. I am unable to discern the process and/or initial source code to do any diagnostics.

```
-- Generating /build/precommit_custom_linux/build/modules/cudaoptflow/CMakeFiles/cuda_compile.dir/src/cuda/./cuda_compile_generated_nvidiaOpticalFlow.cu.o
/usr/local/cuda/bin/nvcc /build/precommit_custom_linux/opencv_contrib/modules/cudaoptflow/src/cuda/nvidiaOpticalFlow.cu -c -o /build/precommit_custom_linux/build/modules/cudaoptflow/CMakeFiles/cuda_compile.dir/src/cuda/./cuda_compile_generated_nvidiaOpticalFlow.cu.o -ccbin /usr/bin/gcc-5 -m64 -D__OPENCV_BUILD=1 -D_USE_MATH_DEFINES -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -D__STDC_FORMAT_MACROS -Xcompiler ,\"-fsigned-char\",\"-W\",\"-Wall\",\"-Werror=return-type\",\"-Werror=non-virtual-dtor\",\"-Werror=address\",\"-Werror=sequence-point\",\"-Wformat\",\"-Werror=format-security\",\"-Winit-self\",\"-Wpointer-arith\",\"-Wuninitialized\",\"-Wno-comment\",\"-fdiagnostics-show-option\",\"-Wno-long-long\",\"-pthread\",\"-fomit-frame-pointer\",\"-ffunction-sections\",\"-fdata-sections\",\"-msse\",\"-msse2\",\"-msse3\",\"-fvisibility=hidden\",\"-Wno-undef\",\"-Wno-missing-declarations\",\"-Wno-shadow\",\"-Wno-strict-aliasing\",\"-Wno-unused-but-set-variable\",\"-O3\",\"-DNDEBUG\",\"-DNDEBUG\" -gencode arch=compute_61,code=sm_61 -D_FORCE_INLINES -Xcompiler -DCVAPI_EXPORTS -Xcompiler -fPIC --std=c++11 -DNVCC -I/usr/local/cuda/include -I/build/precommit_custom_linux/build/3rdparty/ippicv/ippicv_lnx/icv/include -I/build/precommit_custom_linux/build/3rdparty/ippicv/ippicv_lnx/iw/include -I/build/precommit_custom_linux/build -I/usr/include/gdal -I/usr/include/eigen3 -I/usr/include -I/build/precommit_custom_linux/opencv_contrib/modules/cudaoptflow/include -I/build/precommit_custom_linux/build/modules/cudaoptflow -I/build/precommit_custom_linux/opencv_contrib/modules/cudev/include -I/build/precommit_custom_linux/opencv/modules/core/include -I/build/precommit_custom_linux/opencv_contrib/modules/cudaarithm/include -I/build/precommit_custom_linux/opencv/modules/flann/include -I/build/precommit_custom_linux/opencv/modules/imgproc/include -I/build/precommit_custom_linux/opencv_contrib/modules/cudafilters/include -I/build/precommit_custom_linux/opencv_contrib/modules/cudaimgproc/include -I/build/precommit_custom_linux/opencv_contrib/modules/cudawarping/include -I/build/precommit_custom_linux/opencv/modules/dnn/include -I/build/precommit_custom_linux/opencv/modules/features2d/include -I/build/precommit_custom_linux/opencv/modules/imgcodecs/include -I/build/precommit_custom_linux/opencv/modules/calib3d/include -I/build/precommit_custom_linux/opencv/modules/objdetect/include -I/build/precommit_custom_linux/opencv/modules/video/include -I/build/precommit_custom_linux/opencv_contrib/modules/ximgproc/include -I/build/precommit_custom_linux/opencv_contrib/modules/cudalegacy/include -I/build/precommit_custom_linux/opencv_contrib/modules/optflow/include -I/usr/local/cuda/include
/build/precommit_custom_linux/opencv_contrib/modules/cudaoptflow/src/cuda/nvidiaOpticalFlow.cu(76): error: no instance of overloaded function "surf2Dwrite" matches the argument list
            argument types are: (short2, cudaSurfaceObject_t, unsigned long, int, cudaSurfaceBoundaryMode)

1 error detected in the compilation of "/tmp/tmpxft_000010f5_00000000-7_nvidiaOpticalFlow.cpp1.ii".
-- Removing /build/precommit_custom_linux/build/modules/cudaoptflow/CMakeFiles/cuda_compile.dir/src/cuda/./cuda_compile_generated_nvidiaOpticalFlow.cu.o
/usr/bin/cmake -E remove /build/precommit_custom_linux/build/modules/cudaoptflow/CMakeFiles/cuda_compile.dir/src/cuda/./cuda_compile_generated_nvidiaOpticalFlow.cu.o
CMake Error at cuda_compile_generated_nvidiaOpticalFlow.cu.o.cmake:264 (message):
  Error generating file
  /build/precommit_custom_linux/build/modules/cudaoptflow/CMakeFiles/cuda_compile.dir/src/cuda/./cuda_compile_generated_nvidiaOpticalFlow.cu.o


modules/cudaoptflow/CMakeFiles/opencv_cudaoptflow.dir/build.make:84: recipe for target 'modules/cudaoptflow/CMakeFiles/cuda_compile.dir/src/cuda/cuda_compile_generated_nvidiaOpticalFlow.cu.o' failed
make[2]: *** [modules/cudaoptflow/CMakeFiles/cuda_compile.dir/src/cuda/cuda_compile_generated_nvidiaOpticalFlow.cu.o] Error 1
make[2]: Leaving directory '/build/precommit_custom_linux/build'
CMakeFiles/Makefile2:23199: recipe for target 'modules/cudaoptflow/CMakeFiles/opencv_cudaoptflow.dir/all' failed
make[1]: *** [modules/cudaoptflow/CMakeFiles/opencv_cudaoptflow.dir/all] Error 2
```

**Comment by alalek** [comment]
> const unsigned cudaFlags

try `int cudaFlags`

---

> nvidiaOpticalFlow.cu

Problem is from here: https://github.com/opencv/opencv_contrib/pull/2807#issuecomment-761885517
Workarounded through using Ubuntu 18.04 build image (with CUDA 10.0).

**Comment by diablodale** [comment]
That is not possible. The parameter signature is `unsigned int`. It is unsafe to use a signed parameter or to cast it.

I use `unsigned` because the python wrapper infrastructure has a bug if you use `const unsigned int`...it thinks the type is `const unsigned` and the variable is `int xxxx` 😂 a variable name with a space in it. Therefore, I use `const unsigned`.

**Comment by alalek** [comment]
This is not about Python only. Bindings wrappers support only subset of types.
`Unsigned` is not supported due to a multiple reasons. Use 'int'. It is by design.

---

> I can't cast

This doesn't matter until you have such platform in your hands (with CUDA).

**Comment by diablodale** [comment]
I will experiment today, but no guarantees. The most important user to me is C++. And I will not compromise c++ because of python itself or the python infrastructure. The C++ constructor will have `unsigned int`. I will experiment to see if I can somehow have python/infrastructure have instead an `int`. Debugging/enhancing the python wrappers is out-of-scope for this PR -- that is something else for someone else.

I will include a disclaimer in the code/docs that the sign is unreliable and may cause problems now or in the future.
I still suggest we not python wrap this constructor.

**Comment by diablodale** [comment]
I do not see anything that I would change in the constructor.

The two warnings I reported above only appear in 3 platforms. The two test cases I wrote pass and my app itself works well.
If it is desired to not have these warnings, I recommend someone with experience on the python wrapper infrastructure debug this issue.

The python wrapping infrastructure already has some code to support `size_t` and `unsigned int`. Perhaps this code is not working correctly or incomplete.
https://github.com/opencv/opencv/blob/ea41f89b40392fc0c5bdfe0f0c43c4820c4b2eac/modules/python/src2/cv2.cpp#L840-L852

https://github.com/opencv/opencv/blob/ea41f89b40392fc0c5bdfe0f0c43c4820c4b2eac/modules/python/src2/cv2.cpp#L741-L754

I also found at least one function in opencv that uses the size_t support and I don't see warnings about it in the logs.
`CV_WRAP void cv::cuda::DeviceInfo::queryMemory(size_t& totalMemory, size_t& freeMemory) const;`

File: modules/core/include/opencv2/core/cuda.hpp [hunk 2]
```diff
@@ -656,6 +656,18 @@ class CV_EXPORTS_W Stream
...
+
+    @note The cudaFlags parameter is passed unchanged to the underlying api cudaStreamCreateWithFlags() and
+    supports the same parameter values.
+    @code
+        // creates an OpenCV cuda::Stream that manages an asynchronous, non-blocking,
+        // non-default CUDA stream
+        cv::cuda::Stream cvStream(cudaStreamNonBlocking);
+    @endcode
+     */
+    CV_WRAP Stream(const unsigned cudaFlags);
```
Comment (alalek) [inline modules/core/include/opencv2/core/cuda.hpp:669]: Remove CV_WRAP from here to keep this C++ only.

We can add `CV_WRAP static Stream create(int cudaFlags)` or similar for bindings later.

File: modules/core/test/test_mat.cpp [hunk 3]
```diff
@@ -9,6 +9,10 @@
 #include "opencv2/core/eigen.hpp"
 #endif
 
+#if defined(HAVE_CUDA)
+    #include <cuda_runtime.h>
+#endif
```
Comment (alalek) [inline modules/core/test/test_mat.cpp:14]: It make sense to create new file `test_cuda.cpp` for this test

File: modules/core/test/test_mat.cpp [hunk 4]
```diff
@@ -2355,5 +2359,15 @@ TEST(Mat, regression_18473)
     EXPECT_EQ((int)5, (int)m.at<short>(19, 49, 99));
 }
 
+TEST(CUDA_Stream, construct_cudaFlags)
+{
+#ifdef HAVE_CUDA
```
Comment (alalek) [inline modules/core/test/test_mat.cpp:2364]: Guard the whole test (or the file as mentioned above) with `#ifdef HAVE_CUDA`.

IMHO, "#else" branch is not really needed (it is hard enough to introduce such regression).

File: modules/core/include/opencv2/core/cuda.hpp (hunk 2 above)
Comment (diablodale) [reply modules/core/include/opencv2/core/cuda.hpp:669]: I thought of an alternative. I will need to code it and force push to see if it removes the two warnings in the CI platforms. (Functionality has not been a problem...only two warnings on a subset of platforms).

I can code the constructor to accept an `int` (you suggest this is preferred in OpenCV) and also reject negative numbers. By accepting only positive `int` values...I can safely cast it to an `unsigned int`. This alternative supports all the values that the NVidia API currently supports. If in the future NVidia defines a value that is larger than a signed int can represent and someone uses that value, then this constructor will fail. And OpenCV will need to make a choice how to handle NVidia's new value.

I would prefer to not have the cost of a branch/if in the constructor. What is the OpenCV philosophy to use `CV_DbgAssert(param>=0)`. I would prefer that so there is no cost in release builds.
If the core team doesn't like that, then I can use `CV_Assert(param>=0)`.

File: modules/core/test/test_mat.cpp (hunk 3 above)
Comment (diablodale) [reply modules/core/test/test_mat.cpp:14]: ok. I will move it into a new file.

File: modules/core/test/test_mat.cpp (hunk 4 above)
Comment (diablodale) [reply modules/core/test/test_mat.cpp:2364]: Ok. I'll incorporate with the coming PR update.

File: modules/core/src/cuda_stream.cpp [hunk 5]
```diff
@@ -312,6 +313,14 @@ cv::cuda::Stream::Impl::Impl(const Ptr<GpuMat::Allocator>& allocator) : stream(0
     ownStream = true;
 }
 
+cv::cuda::Stream::Impl::Impl(const int cudaFlags) : stream(0), ownStream(false)
+{
+    CV_DbgAssert(cudaFlags >= 0);
```
Comment (alalek) [inline modules/core/src/cuda_stream.cpp:318]: IMHO, this check doesn't help.

My suggestion is to remove that.

File: modules/python/test/test_cuda.py [hunk 6]
```diff
@@ -33,6 +33,8 @@ def test_cuda_interop(self):
         self.assertTrue(cuMat.cudaPtr() != 0)
         stream = cv.cuda_Stream()
         self.assertTrue(stream.cudaPtr() != 0)
+        asyncstream = cv.cuda_Stream(1)
```
Comment (alalek) [inline modules/python/test/test_cuda.py:36]: Consider adding a comment with description for magic numbers:

```
asyncstream = cv.cuda_Stream(1)  # cudaStreamNonBlocking
```

**Review by alalek — COMMENTED** [review]
What is wrong with `cv::cuda::StreamAccessor`?
It was specially added to work with CUDA SDK directly (see all CUDA enums / types, respect type safety).

`Stream` API has limitation, which doesn't allow to use 3rdparty dependencies (CUDA SDK).

---

`StreamAccessor` is not a part of bindings.

File: modules/python/test/test_cuda.py (hunk 6 above)
Comment (diablodale) [reply modules/python/test/test_cuda.py:36]: The CUDA api defines are documented in the doxygen for the constructor. I can also add it to this test file, no problem.

File: modules/core/src/cuda_stream.cpp (hunk 5 above)
Comment (diablodale) [reply modules/core/src/cuda_stream.cpp:318]: I do believe we need a limit test somewhere. It is illegal to pass negative numbers. Since the python shim infrastructure has a bug/issue (gives non-important warnings), I can't use unsigned int. This was discussed multiple times before.

**Comment by diablodale** [comment]
Did you write again "What is wrong with cv::cuda::StreamAccessor?"?
If you did, this was discussed multiples times before.

File: modules/core/include/opencv2/core/cuda.hpp [hunk 7]
```diff
@@ -656,6 +656,18 @@ class CV_EXPORTS_W Stream
...
+
+    @note The cudaFlags parameter must be a positive integer; it is passed to the underlying
+    api cudaStreamCreateWithFlags() and supports the same parameter values.
+    @code
+        // creates an OpenCV cuda::Stream that manages an asynchronous, non-blocking,
+        // non-default CUDA stream
+        cv::cuda::Stream cvStream(cudaStreamNonBlocking);
+    @endcode
+     */
+    CV_WRAP Stream(const int cudaFlags);
```
Comment (alalek) [inline modules/core/include/opencv2/core/cuda.hpp:669]: I took a look a bit more at bindings generators.
There is `size_t` type support in bindings.

File: modules/core/src/cuda_stream.cpp [hunk 8]
```diff
@@ -293,6 +293,7 @@ class cv::cuda::Stream::Impl
 
     Impl();
     Impl(const Ptr<GpuMat::Allocator>& allocator);
+    Impl(const int cudaFlags);
```
Comment (alalek) [inline modules/core/src/cuda_stream.cpp:296]: Please use type safe `unsigned int` in internal implementation.

File: modules/core/include/opencv2/core/cuda.hpp (hunk 7 above)
Comment (diablodale) [reply modules/core/include/opencv2/core/cuda.hpp:669]: I am pushing again to check for any warnings on the CI platforms. Using `size_t` in this way is not in alignment with intention of the type https://en.cppreference.com/w/cpp/types/size_t
It may work on all common platforms/data models today. And it may fail in the future due to using `size_t` in a way the C++ standard did not intend.

File: modules/core/src/cuda_stream.cpp (hunk 8 above)
Comment (diablodale) [reply modules/core/src/cuda_stream.cpp]: done

**Review by alalek — APPROVED** [review]
Thank you :+1:

**Review by asmorkalov — APPROVED** [review]
